File: cpp/mpi_pi_lowgran.c
// This program is to caculate PI using MPI and numerical integration
// If f(x) = 4 * sqrt(1 - x^2), then PI is the intergral of f(x) from 0 to 1

#include <cmath>
#include <mpi.h>

constexpr double f(double x)
{
    return 4 * sqrt(1 - x * x);
}

int main (int argc, char* argv[])
{
    constexpr int N = 1e7;
    constexpr double from = 0.;
    constexpr double to = 1.;
    
    constexpr double d = (to - from) / static_cast<double>(N);

    // Initialize MPI
    MPI_Init(&argc, &argv);

    // Get process id and total number of processes size
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);    
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    
    // Wait until all processes arrive here, then continue
    MPI_Barrier(MPI_COMM_WORLD);
    const double begin = MPI_Wtime();
    
    //Each process caculates a part of the sum. 
    double subsum = 0, sum = 0, total_sum;
    for (int i = rank; i < N; i += size)
    {
        const double x = from + i * d;
        const double fx = f(x);
        subsum = fx * d;
        MPI_Reduce(&subsum, &sum, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);
        total_sum += sum;
    }
    //std::cout << "proc. " << rank << " finished" << std::endl;
    
    //Sum up all sub sums
    
    //Synchronize all processes and get the end time
    MPI_Barrier(MPI_COMM_WORLD);
    const double end = MPI_Wtime();
    
     //Caculate and print PI
    if (!rank)
        std::cout << "Processes = " << size << ", Exec. duration = " << end - begin << "s, PI = " << total_sum << std::endl;
    
    
    MPI_Finalize();
    
    return 0;
}

File: python/mpi_pi_lowgran.py
#!/usr/bin/env python 

# Caculate PI using MPI and numerical integration
# of a quarter circle

from mpi4py import MPI
import math

def f(x: float) -> float:
	return math.sqrt(1 - x ** 2)

N: int = int(1e6)

fromV: float = 0.
toV: float = 1.

d: float = (toV - fromV) / N

# Initialize MPI
comm = MPI.COMM_WORLD
size: int = comm.Get_size()
rank: int = comm.Get_rank()

# Wait until all processes arrive here, then continue
comm.Barrier()
begin: float = MPI.Wtime()

sum: float = 0
# Each process caculates a part of the sum. 
for i in range(rank, N, size):
    x: float = fromV + i * d
    fx: float = f(x)
    subsum: float = fx * d
    reduced = comm.reduce(subsum, op = MPI.SUM, root=0)
    if rank == 0:
        sum += reduced

# Sum up all sub sums

# Synchronize all processes and get the end time
comm.Barrier()
end: float = MPI.Wtime()

# Caculate and print PI
if rank == 0:
	print("Processes = ", size, ", Exec. duration = ", end - begin, "s, PI = ", 4 * sum, ", error = ", 4 * sum - math.pi)


MPI.Finalize()